### ISSUE TITLE
Client crash on RabbitMQ connection failure due to temporary name resolution failure

### ISSUE THREAD

**Issue by rmc3** [sensu/sensu#747]
Running 0.12.6, the Sensu client crashes with an EventMachine connection error on (re)connecting to RabbitMQ when name resolution is not available.

```
/opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/eventmachine.rb:776:in `connect_server': unable to resolve server address (EventMachine::ConnectionError)
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/eventmachine.rb:776:in `reconnect'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/amq-client-1.0.2/lib/amq/client/async/adapters/event_machine.rb:95:in `block in periodically_reconnect'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/em/timers.rb:56:in `call'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/em/timers.rb:56:in `fire'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `call'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run_machine'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/sensu-0.12.6/lib/sensu/client.rb:12:in `run'
        from /opt/contegix/sensu/embedded/lib/ruby/gems/1.9.1/gems/sensu-0.12.6/bin/sensu-client:10:in `<top (required)>'
        from /opt/contegix/sensu/bin/sensu-client:23:in `load'
        from /opt/contegix/sensu/bin/sensu-client:23:in `<main>'
```

I would expect this error to instead be caught and the connection retried later. A common issue we see causing this is a temporary network failure on the network providing both name resolution and the RabbitMQ connection. This results in disconnection from RabbitMQ and the subsequent crash of the client when it fails to look up the RabbitMQ server while reconnecting.

**Comment by iiro** [comment]
Hi,

having still problems with this...

Is this wanted behaviour that eg. Sensu client restart fails if the Sensu host is not found in DNS?

/opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:664:in `connect_server': unable to resolve server address (EventMachine::ConnectionError)
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:664:in `bind_connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:640:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/amqp-1.5.0/lib/amqp/session.rb:457:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/amqp-1.5.0/lib/amqp.rb:242:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-2.4.0/lib/sensu/transport/rabbitmq.rb:118:in `setup_connection'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-2.4.0/lib/sensu/transport/rabbitmq.rb:147:in `connect_with_eligible_options'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-2.4.0/lib/sensu/transport/rabbitmq.rb:14:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-2.4.0/lib/sensu/transport.rb:22:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.18.1/lib/sensu/daemon.rb:184:in `setup_transport'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.18.1/lib/sensu/client/process.rb:390:in `start'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.18.1/lib/sensu/client/process.rb:19:in `block in run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `call'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run_machine'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.18.1/lib/sensu/client/process.rb:18:in `run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.18.1/bin/sensu-client:10:in `<top (required)>'
	from /opt/sensu/bin/sensu-client:23:in `load'
	from /opt/sensu/bin/sensu-client:23:in `<main>'

It cannot continue retrying on background in some intervals?

**Comment by malpania** [comment]
We are encountering the same problem using : sensu-0.20.3-1.x86_64
Our client nodes sometimes start before the sensu server is up and fail to resolve the DNS hostname. However the client is unable to reconnect once the sensu server comes back up. 

/etc/sensu/conf.d/config.json 
{
  "rabbitmq": {
    "host": "sensu.server",
    "port": 5671,
...

/opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:664:in `connect_server': unable to resolve server address
 (EventMachine::ConnectionError)
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:664:in `bind_connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:640:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/amqp-1.5.0/lib/amqp/session.rb:457:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/amqp-1.5.0/lib/amqp.rb:242:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-3.2.0/lib/sensu/transport/rabbitmq.rb:119:in `setup_connection'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-3.2.0/lib/sensu/transport/rabbitmq.rb:148:in `connect_with_eligible
_options'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-3.2.0/lib/sensu/transport/rabbitmq.rb:15:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-transport-3.2.0/lib/sensu/transport.rb:22:in `connect'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.20.3/lib/sensu/daemon.rb:187:in `setup_transport'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.20.3/lib/sensu/client/process.rb:412:in `start'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.20.3/lib/sensu/client/process.rb:19:in `block in run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `call'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run_machine'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/eventmachine-1.0.3/lib/eventmachine.rb:187:in `run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.20.3/lib/sensu/client/process.rb:18:in `run'
	from /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-0.20.3/bin/sensu-client:10:in `<top (required)>'
	from /opt/sensu/bin/sensu-client:23:in `load'
	from /opt/sensu/bin/sensu-client:23:in `<main>'

**Comment by ball-hayden** [comment]
Also seeing this problem still.

sensu-client 0.22.1

Similar backtrace.

**Comment by calebhailey** [comment]
@ball-hayden could you try to reproduce this with 0.23? I wonder if the new improvements to Sensu transport will help resolve this issue; see: https://sensuapp.org/docs/0.23/changelog

**Comment by ball-hayden** [comment]
@calebhailey - looks like it has indeed been fixed by 0.23.
Thank you.

**Comment by calebhailey** [comment]
@ball-hayden thanks for confirming! I'll go ahead and close this now! \o/
